src/ngraph/ngraph_imperative.cc

```diff
@@ -87,69 +87,128 @@ void NGImperative::parse_ngraph() {
     }
 }
 
+bool compute_forward_imperative(const nnvm::NodeAttrs &attrs,
+                                const mxnet::OpContext &ctx,
+                                const std::vector<mxnet::NDArray> &inputs,
+                                const std::vector<mxnet::OpReqType> &req,
+                                const std::vector<mxnet::NDArray> &outputs) {
+  std::shared_ptr<Graph> op_ng;
+  int mode = static_cast<int>(GraphExeMode::kInfer);
+  if (!sparse_check(inputs) && !sparse_check(outputs)) {
+    // thread local cache for ngraph op
+    static thread_local NGIOpCache ngicache;
+    auto op_key = get_ngiop_key(attrs, ctx.run_ctx.ctx, inputs);
+    op_ng = ngicache[op_key];
+    if (!op_ng) {
+      NGImperative ngi(attrs, ctx.run_ctx.ctx, inputs, &req, outputs);
+      op_ng = ngicache[op_key] = ngi.get_op_ngraph();
+      if (ngraph_log_verbose && op_ng)
+        LOG(INFO) << "Caching... " << attrs.op->name;
+    }
+    if (ctx.is_train) {
+      mode = static_cast<int>(GraphExeMode::kTrain);
```

Comment:
Near the top of this function you set `mode` to `kInfer`, and then in this line you potentially change it.

Would it have been possible to just set `mode` once at the top of the function?

Reply:
was done by different team member in different PR, but will modify the mode, it will be more clear.